### PR TITLE
Protect data & co.

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
- #16 protect data (State)
- #15 encapsulation (fix parameter.date support)
- #14 early variable binding
- #10 easy charts and tables ```<{[26,74,65]}|chart|type=pie|>```
- #18 date_selector -> date
- #4 vertical indicator ```<|MAE: {mae}|indicator|value={val}|min=20|max=-20|orientation=vertical|>```

### CODE REVIEW

File: gui/src/components/Taipy/DateSelector.tsx [hunk 1]
```diff
@@ -63,22 +70,37 @@ const DateSelector = (props: DateSelectorProps) => {
...
-        />
+    return (
+        <Box id={id} className={className} sx={boxSx}>
+            {editable ? (
+                withTime ? (
+                    <DateTimePicker
+                        value={value}
+                        onChange={handleChange}
+                        renderInput={renderInput}
+                        className={getSuffixedClassNames(className, "-picker")}
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/DateSelector.tsx:81]: The code is used three times. Better to factorize?
And... not too sure about the 'picker' part...

File: gui/src/components/Taipy/DateSelector.tsx [hunk 2]
```diff
@@ -63,22 +70,37 @@ const DateSelector = (props: DateSelectorProps) => {
...
+                        disabled={!active}
+                    />
+                )
+            ) : (
+                <Field
+                    dataType="datetime.datetime"
+                    defaultValue={props.defaultDate}
+                    value={props.date}
+                    format={props.format}
+                    id={id + "-field"}
```
Comment (FabienLelaquais) [inline gui/src/components/Taipy/DateSelector.tsx:99]: Why not use 'id' at this point? The input field is not rendered...

File: tests/taipy/gui/control/test_date_selector.py [hunk 3]
```diff
@@ -3,29 +3,29 @@
 from taipy.gui import Gui
 
 
-def test_date_selector_md_1(gui: Gui, helpers):
+def test_date_md_1(gui: Gui, helpers):
```
Comment (FabienLelaquais) [inline tests/taipy/gui/control/test_date_selector.py:6]: You may even want to rename the file itself at this point...

File: taipy/gui/utils/_state.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import typing as t
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_state.py:1]: That really looks great. Eager to see that in action!
Should the Gui object know about those variables though?

File: taipy/gui/data/array_dict_data_accessor.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    __types = (dict, list, tuple)
+
+    @staticmethod
+    def get_supported_classes() -> t.List[str]:
+        return [t.__name__ for t in ArrayDictDataAccessor.__types]
+
+    def __get_dataframe(self, value: t.Any) -> pd.DataFrame:
+        if len(value):
+            if isinstance(value[0], (str, int, float, bool)):
```
Comment (FabienLelaquais) [inline taipy/gui/data/array_dict_data_accessor.py:19]: Shouldn't we test is all values are kind of the same type?
Or maybe the DataFrame constructor does it?

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Excellent, excellent job.
A few remarks that might need tiny changes...

File: gui/src/components/Taipy/DateSelector.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/DateSelector.tsx:81]: Wrong copy/paste, it should only be used 2 times with the same suffix

Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/DateSelector.tsx:81]: And in different cases

File: gui/src/components/Taipy/DateSelector.tsx (hunk 2 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/DateSelector.tsx:99]: Id is used at the parent level

File: tests/taipy/gui/control/test_date_selector.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply tests/taipy/gui/control/test_date_selector.py:6]: Will do

File: taipy/gui/utils/_state.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_state.py:1]: ?
It does

File: taipy/gui/data/array_dict_data_accessor.py (hunk 5 above)
Comment (FredLL-Avaiga) [reply taipy/gui/data/array_dict_data_accessor.py:19]: I guess dataframe will throw an error if it's not happy 😁

File: gui/src/components/Taipy/DateSelector.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/DateSelector.tsx:81]: Do you have another proposal for picker?

File: tests/taipy/gui/data/test_arry_dict_data_accessor.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import pytest
```
Comment (dinhlongviolin1) [inline tests/taipy/gui/data/test_arry_dict_data_accessor.py]: File name typo

File: taipy/gui/gui.py [hunk 7]
```diff
@@ -280,9 +284,11 @@ def __update_var(self, var_name: str, value: t.Any, propagate=True, holder: Taip
         elif holder:
             modified_vars.update(self._evaluate_holders(hash_expr))
         # TODO: what if _update_function changes 'var_name'... infinite loop?
-        if self.on_change:
+        if hasattr(self, "on_change") and isinstance(self.on_change, FunctionType):
```
Comment (dinhlongviolin1) [inline taipy/gui/gui.py:287]: Thanks

File: taipy/gui/gui.py [hunk 8]
```diff
@@ -519,32 +525,31 @@ def __on_action(self, id: t.Optional[str], payload: t.Any) -> None:
         if action and hasattr(self, action):
             if self.__call_function_with_args(action_function=getattr(self, action), id=id, payload=payload):
                 return
-        if self.on_action:
+        if hasattr(self, "on_action") and isinstance(self.on_action, FunctionType):
```
Comment (dinhlongviolin1) [inline taipy/gui/gui.py:528]: Thanks again

**Review by dinhlongviolin1 — APPROVED** [review]
Great job! I just have 1 question. Since gui has been replaced by state, from what I understand, the users wont be able to access gui public method through state. Correct me if I am wrong

File: tests/taipy/gui/data/test_arry_dict_data_accessor.py (hunk 6 above)
Comment (FredLL-Avaiga) [reply tests/taipy/gui/data/test_arry_dict_data_accessor.py]: Fixed

**Comment by FredLL-Avaiga** [comment]
> Great job! I just have 1 question. Since gui has been replaced by state, from what I understand, the users wont be able to access gui public method through state. Correct me if I am wrong
That's true and it's "by design":
- the user variable are accessed via State
- the Gui behavior via Gui

**Comment by dinhlongviolin1** [comment]
> > Great job! I just have 1 question. Since gui has been replaced by state, from what I understand, the users wont be able to access gui public method through state. Correct me if I am wrong
> > That's true and it's "by design":
> 
> * the user variable are accessed via State
> * the Gui behavior via Gui

From what I have seen in the code, users can't access gui directly from the parameters anymore right

**Comment by FredLL-Avaiga** [comment]
> From what I have seen in the code, users can't access gui directly from the parameters anymore right

they should have an gui instance globally accessible ;-)

**Comment by dinhlongviolin1** [comment]
> > From what I have seen in the code, users can't access gui directly from the parameters anymore right
> 
> they should have an gui instance globally accessible ;-)

That is exactly what I am expecting. If Fabien's ok with it, I am

**Review by FabienLelaquais — APPROVED** [review]
👍

**Pull request merged**